core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageFeatureStep.java

```diff
@@ -538,108 +544,111 @@ public void write(String s, byte[] bytes) {
-            MethodCreator mv = file.getMethodCreator("registerClass" + count++, void.class, Feature.BeforeAnalysisAccess.class);
-            mv.setModifiers(Modifier.PRIVATE | Modifier.STATIC);
-            registerForReflection.invokeStaticMethod(mv.getMethodDescriptor(), registerForReflection.getMethodParam(0));
-
-            TryBlock tc = mv.tryBlock();
-
-            ResultHandle clazz = tc.loadClassFromTCCL(entry.getKey());
-            //we call these methods first, so if they are going to throw an exception it happens before anything has been registered
-            ResultHandle constructors = tc
-                    .invokeVirtualMethod(ofMethod(Class.class, "getDeclaredConstructors", Constructor[].class), clazz);
-            ResultHandle methods = tc.invokeVirtualMethod(ofMethod(Class.class, "getDeclaredMethods", Method[].class), clazz);
-            ResultHandle fields = tc.invokeVirtualMethod(ofMethod(Class.class, "getDeclaredFields", Field[].class), clazz);
-
-            if (!entry.getValue().weak) {
-                ResultHandle carray = tc.newArray(Class.class, tc.load(1));
-                tc.writeArrayValue(carray, 0, clazz);
-                tc.invokeStaticMethod(ofMethod(RUNTIME_REFLECTION, "register", void.class, Class[].class),
-                        carray);
-
-                if (entry.getValue().constructors) {
-                    tc.invokeStaticMethod(
-                            ofMethod(RUNTIME_REFLECTION, "register", void.class, Executable[].class),
-                            constructors);
-                } else if (!entry.getValue().ctorSet.isEmpty()) {
-                    ResultHandle farray = tc.newArray(Constructor.class, tc.load(1));
+            // To avoid getting MethodTooLargeException, the methods only manage up to CLASSES_TO_REGISTER_BATCH_SIZE entries
+            // To avoid getting ClassTooLargeException, the methods need to manage several classes
+            if (index++ % CLASSES_TO_REGISTER_BATCH_SIZE == 0) {
+                if (currentRegisterClass != null) {
+                    currentRegisterClass.returnVoid();
+                }
+                currentRegisterClass = file.getMethodCreator("registerClasses" + count++, void.class,
+                        Feature.BeforeAnalysisAccess.class);
+                currentRegisterClass.setModifiers(Modifier.PRIVATE | Modifier.STATIC);
+            }
+            boolean hasConstructorSet = !entry.getValue().weak && !entry.getValue().constructors
+                    && !entry.getValue().ctorSet.isEmpty();
+            boolean hasMethodSet = !entry.getValue().weak && !entry.getValue().methods && !entry.getValue().methodSet.isEmpty();
+            boolean hasFieldSet = !entry.getValue().weak && !entry.getValue().fields && !entry.getValue().fieldSet.isEmpty();
+            boolean hasSerialization = entry.getValue().serialization;
+            boolean tryBlock = hasConstructorSet || hasMethodSet || hasFieldSet || hasSerialization;
```

Comment:
If I am not mistaken that should be:

```suggestion
            boolean hasConstructorSet = !entry.getValue().constructors
                    && !entry.getValue().ctorSet.isEmpty();
            boolean hasMethodSet = !entry.getValue().weak && !entry.getValue().methods && !entry.getValue().methodSet.isEmpty();
            boolean hasFieldSet = !entry.getValue().weak && !entry.getValue().fields && !entry.getValue().fieldSet.isEmpty();
            boolean hasSerialization = entry.getValue().serialization;
            boolean tryBlock = !entry.getValue().weak && (hasConstructorSet || hasMethodSet || hasFieldSet || hasSerialization);
```

Reply:
I would rather be `boolean tryBlock = (!entry.getValue().weak && (hasConstructorSet || hasMethodSet || hasFieldSet)) || hasSerialization;` which doesn't sound to be easier to read, I would rather prefer to rename the variables if you prefer to something like `hasConstructorsToHandle`, WDYT?

Reply:
Sounds good to me. Thanks